File: algorithm/python/AllNodesDistanceKInBinaryTree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def distanceK(self, root, target, k):
        parent = dict()
        res = []
        def getParent(node):
            if node.left:
                parent[node.left.val] = node
                getParent(node.left)
            if node.right:
                parent[node.right.val] = node
                getParent(node.right)

        getParent(root)
        print(parent)
        def getRes(node, pre, depth):
            if not node: return
            
            if depth == k:
                res.append(node.val)
                return
            if node.left != pre:
                getRes(node.left, node, depth+1)
            if node.right != pre:
                getRes(node.right, node, depth+1)
            if node.val not in parent.keys() and parent[node.val] != pre:
                getRes(parent[node.val], node, depth+1)
        
        getRes(target, None, 0)
        return res

File: algorithm/JavaScript/AllNodesDistanceKInBinaryTree.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */
/**
 * @param {TreeNode} root
 * @param {TreeNode} target
 * @param {number} k
 * @return {number[]}
 */
 var distanceK = function(root, target, k) {
    const parent = new Map()
    const res = []
    const getParent = (node) => {
      if (node.left !== null) {
        parent.set(node.left.val, node);
        getParent(node.left)
      }
      if (node.right !== null) {
        parent.set(node.right.val, node)
        getParent(node.right)
      }
    }
    getParent(root)

    const findRes = (node, from, depth, k) => {
      if (node == null) return

      if (depth === k) {
        res.push(node.val)
        return
      }

      if (node.left !== from) {
        findRes(node.left, node, depth+1, k)
      }
      if (node.right !== from) {
        findRes(node.right, node, depth+1, k)
      }
      if (parent.get(node.val) !== from) {
        findRes(parent.get(node.val), node, depth+1, k)
      }
    }

    findRes(target, null, 0, k)
    return res

};